### PR TITLE
Wireguard component

### PR DESCRIPTION
**Pull request by lhoracek**
# What does this implement/fix?

This is an implementation of [WireGuard®](https://www.wireguard.com/) VPN for [ESPHome](https://esphome.io/), based on [Wireguard Implementation for ESP-IDF](https://github.com/trombik/esp_wireguard) (by [@trombik](https://github.com/trombik)).
Using @droscy's fork (https://github.com/droscy/esp_wireguard) that allows state to be watched.

TODO:
  - [x] Connected binary sensor
  - [x] Last handshake sensor
  - [x] Allowed IPs
  - [x] Tests

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** [esphome/esphome-docs#2948](https://github.com/esphome/esphome-docs/pull/2948)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
wireguard:
  address: x.y.z.w
  private_key: private_key=
  peer_endpoint: wg.server.example
  peer_public_key: public_key=
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by antonio-fiol** [comment]
I tested this PR, successfully. Is there any way I can help so that it becomes OK to merge?

**Comment by lhoracek** [comment]
> I tested this PR, successfully. Is there any way I can help so that it becomes OK to merge?

I'm trying to find time to extend it with those two sensors, that would help! I could move to adding tests. 🙇

**Comment by antonio-fiol** [comment]
> I'm trying to find time to extend it with those two sensors

Happy to (try to) help, but... What two sensors? It seems that I am missing some context.

**Comment by lhoracek** [comment]
> > I'm trying to find time to extend it with those two sensors
> 
> Happy to (try to) help, but... What two sensors? It seems that I am missing some context.

I think this component needs two sensors mentioned in the TODO in PR description: 
 - Connection status binary sensor
 - Last handshake timestamp sensor (this will need a change in the library, as this value is encapsulated and not accessible straight away :-/, but making just another fork of the lib will be quite easy, and I'll do that this week)

Ideally it waits for time component to get initialised (on_time) as currently it does this quite optimistically by having late startup priority.

**Comment by antonio-fiol** [comment]
I'll take a look and see if I can help on these items.

**Comment by lightingman117** [comment]
Hi there, thank you for working on this!

Really excited to have an official WG component in ESPHome :)

Can't wait to have some smartplugs at work :)
[Though, 8266 support would be necessary due to most smartplugs running 8266]

---

I am not a dev, so forgive me if already implemented.
(FWIW, I did read the commits, but I can't tell if this is included or not and simply not documented, yet).

 Shouldn't the WG_Client have:
 1) `peer_endpoint_port` so that users can change from default (51820)
 2) `peer_preshared_key` optional if desired

Additionally, 'latest handshake' is a good indicator of VPN connection; but not that data is flowing. A better metric is the received data amount it will say a few bits traversed the VPN until real data when it will quickly get into Kib.

Given the WG Win10 GUI [amongst others] shouldn't the `wireguard.h` file be a bit more consistent?

```
[Interface]
PrivateKey = private_key=
Address = 10.0.0.10/32

[Peer]
PublicKey = public_key=
PresharedKey = preshared_key=
AllowedIPs = 10.0.0.1/32, 192.168.1.0/24
Endpoint = smarthome.duckdns.org:51820
```

Thus `wireguard.h` would look more like this:
```
wireguard:
  #interface
  address: "10.0.0.10/32"
  private_key: private_key=

  #peer
  peer_public_key: public_key=
  peer_preshared_key: preshared_key=
  peer_allowed_ip: "10.0.0.1/32, 192.168.1.0/24"
  peer_endpoint: wg.server.example
  peer_port: 51820
```

**Comment by jpeletier** [comment]
I tested this and worked right out of the box. Awesome job @lhoracek @droscy ... Will work beautifully to manage a remote ESPHome dealing with a boiler and lighting and integrate it with the others as if it was in the same network... 

Is there a backlog of stuff somewhere for this component?, please let me know so I can help.

**Comment by droscy** [comment]
Hi @jpeletier thanks for the feedback.

> Is there a backlog of stuff somewhere for this component?, please let me know so I can help.

I'm currently working on the `allowed_ips` list and that should almost complete the original TODOs. Tests and documentation remain to be done... always the last :-D (also for `esp_wireguard` library).

I don't have other features to be added but feel free to review the code, improvements are always welcome.

**Comment by droscy** [comment]
@lhoracek could you please update the description of this PR with new library reference, new examples, etc. Thanks.

**Comment by bgamari** [comment]
When I attempted to use this I found that a Wireguard connection would not be initiated until the MQTT client had connected. This is problematic if you expect to connect to the MQTT broker via VPN. Perhaps this was user error?

**Comment by bgamari** [comment]
> When I attempted to use this I found that a Wireguard connection would not be initiated until the MQTT client had connected. This is problematic if you expect to connect to the MQTT broker via VPN. Perhaps this was user error?

I did a bit of digging here. The reason is that the component's `setup_priority` is set to `LATE`, which is well after that of `mqtt_client` (`AFTER_WIFI`). As `mqtt_client` prevents the setup phase from proceeding (via `Component::can_proceed`) until a connection has been established, setup stalls.

I can think of two ways of fixing this:

 * raise the priority of the `wireguard` component to a new `setup_priority::VPN` (which would sit between `AFTER_WIFI` and `BEFORE_CONNECTION`)
 * make `mqtt_client`'s `can_proceed` logic configurable

My sense is that the first option is preferable; it seems wrong for VPN setup to happen late in setup.

**Comment by droscy** [comment]
> This is problematic if you expect to connect to the MQTT broker via VPN

Absolutely, I agree with you.

The latest code of wireguard component (not yet merged here, see links below) already has the priority raised to `AFTER_WIFI`. 
 - https://github.com/esphome/feature-requests/issues/1444#issuecomment-1556090095
 - lhoracek/esphome#9 or lhoracek/esphome#7

@bgamari do you know how we can "create" the new `VPN` level? I think this could be a good solution. Can it be done directly inside this component or should it be a review of the ESPHome core?

**Comment by antonio-fiol** [comment]
@droscy  see https://github.com/esphome/esphome/blob/6ecf4ecac6125219253672e734d7f7f14bec9b11/esphome/core/component.cpp#L13

I understand that the component itself (wireguard.h line 17) could return any float, like 240.0f to go right after wifi... but that would be quite ugly and probably it should only be done for testing purposes.

**Comment by bgamari** [comment]
@droscy, see https://github.com/bgamari/esphome/commit/ec905b633c1fbcc755d27397d2c8a37043d58f43. Feel free to take that commit if you would like. I am using it locally.

**Comment by lhoracek** [comment]
Should we take the time_sync into account when initialising or are we leaving for the component to retry? I ignored the requirement to have time synced initially.
But speeding up the initialisation should solve couple issues I'm seeing when trying to do stuff over VPN, like updating the Nextion firmware in NSPanels connected over VPN.

**Comment by droscy** [comment]
Sorry @lhoracek but I don't understand what you asked.

The setup step consists in `esp_wireguard_init()` and `esp_wireguard_connect()`: the first is always executed because it doesn't require the clock to be synchronization, while the second requires the synchronization. The `setup()` tries to execute the connect because the clock, in the mean time, could have already been synchronized, and there is no need to wait the second synchronization. If it fails the callback will start the connection as soon as the clock is synchronized.

> But speeding up the initialisation should solve couple issues I'm seeing when trying to do stuff over VPN, like updating the Nextion firmware in NSPanels connected over VPN.

What kind of issues?

To speed up the initialization probably we should pick https://github.com/bgamari/esphome/commit/ec905b633c1fbcc755d27397d2c8a37043d58f43 commit but it changes the `AFTER_WIFI` value affecting every component that is using it and I don't know if it can be done here... we can create the new `setup_priority::VPN` only, but at least the sntp component should be edited.

**Comment by khenderick** [comment]
I'm currently running `https://github.com/droscy/esphome` ref `wireguard/main`. Is there anything I can do to further help testing this particular PR to help eventually having this as native component?

**Comment by thomas0bernard** [comment]
Hello! Thanks a lot @lhoracek and @droscy for the awesome work. I've spent the past few days working with the component and I think a couple things are worth raising to your attention.

**First**, either my understanding of @droscy 's addition of `peer_allowed_ips` is wrong, or it's not behaving correctly. My use case connects to a WG peer in the 10.44.0.0/16 subnet, then connects to an MQTT broker in the 10.0.0.0/16 subnet. I had first set up the component in the following way:
```
wireguard:
  address: 10.44.100.100
  netmask: 255.255.0.0
  peer_endpoint: ***.com
  private_key: W***w=
  peer_public_key: 9***Y=
  peer_allowed_ips:
    - 10.0.0.0/16
```
That didn't work, nor did removing `peer_allowed_ips` altogether, which if I understood correctly should allow 0.0.0.0/0, i.e. pass everything through WG. While digging through the commit history, I found that the previous implementation determined the addresses to "pass" through the WG interface with the `netmask` parameter, and found that setting it to 0.0.0.0 instead of 255.255.0.0 worked and the device connected to my MQTT broker. I spend the past few hours digging to try and understand what's happening here to no great success; the allowed IPs are added correctly as far as I can tell. I ended up adding traces in `wireguardif_output()` (wireguardif.c:197), where it looks like the lookup against the allowed IPs happen, but it's not even getting called with the netmask set to 255.255.0.0 (it does with 0.0.0.0).

Anyhow, my use-case works with setting the netmask to 0.0.0.0, but I can imagine scenarios where a user would want more granular control on which addresses to pass through so I'm still curious if there's a fix to this.

**Second**, arguably more of an edge case. The DNS resolution that happens in `esp_wireguard_peer_init()` can take long enough on some network setups to trigger the task watchdog timeout panic. I saw that esphome's OTA component seems to have dealt with the same issue, temporarily raising the timeout value for a long operation and bringing it back down after. I'm wondering if this is too much of an edge case to get an "official" fix or if it's worth looking into.

Thank you so much!

**Comment by droscy** [comment]
> I've spent the past few days working with the component and I think a couple things are worth raising to your attention.

Thanks for your feedback @thomas0bernard, I've opened the two issues above in my esphome repo just to don't forget them and to have dedicated discussions. I'll work on them.

> I'm currently running `https://github.com/droscy/esphome` ref `wireguard/main`. Is there anything I can do to further help testing this particular PR to help eventually having this as native component?

Any test you can perform @khenderick is welcome. The more we discover the better the code will be. Also in the documentation.

@lhoracek could you please link the PR of the documentation on top description of this PR?

**Comment by lhoracek** [comment]
🔥 Big update by @droscy and @thomas0bernard who made this thing mature and ready for upstream. 
🧪 Ready for testing.
❤️ All contributions are appreciated!

**Review by CarlosGS — APPROVED** [review]
Looking great, maybe I'd suggest, optionally:
- Forking the adapted wireguard repository under the ESPHome organization
- In the documentation, more explicitly say "Currently only supported by ESP32 platform"

This is a powerful contribution, thank you very much! :-)

**Comment by droscy** [comment]
> In the documentation, more explicitly say "Currently only supported by ESP32 platform"

Sure, I'll add it as soon as I update the doc with these links https://github.com/esphome/feature-requests/issues/1444#issuecomment-1579980849

> Forking the adapted wireguard repository under the ESPHome organization

Do you mean merging the underlying [library](https://github.com/droscy/esp_wireguard) files inside ESPHome code? I have a draft branch for it because I already tried but I ended in endless compilation errors and I gave up. Maybe we will try again...

> This is a powerful contribution, thank you very much! :-)

Happy to contribute ;-)

**Comment by CarlosGS** [comment]
For the fork, I meant to say that ESPHome has repositories for other dependencies (i.e. https://github.com/esphome/noise-c or https://github.com/esphome/libsodium-esphome), so maybe it would be good to keep the adapted Wireguard forked within ESPHome's repositories too?
But I'm not sure how important it is really, @jesserockz what do you think?

**Comment by droscy** [comment]
> For the fork, I meant to say that ESPHome has repositories for other dependencies [...], so maybe it would be good to keep the adapted Wireguard forked within ESPHome's repositories too?

Yes, it would be good, so the `esp_wireguard` library will not belong to my github user anymore.

**Comment by droscy** [comment]
After merging latest `dev` I cannot compile wireguard anymore due to #4926. I've changed my code as described in that PR but the compilation always fails with

```
src/esphome/components/wireguard/wireguard.cpp:10:10: fatal error: esphome/core/time.h: No such file or directory
 #include "esphome/core/time.h"
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
*** [.pioenvs/esp07/src/esphome/components/wireguard/wireguard.cpp.o] Error 1
```

@jesserockz, what can I do?

**Comment by CarlosGS** [comment]
@droscy I've checked and all seems fine with a fresh install of the ESPHome (dev) plugin, using your updated branch:
```yaml
external_components:
  - source: github://droscy/esphome@wireguard/main
    components:
      - wireguard
      - wireguard_status
      - wireguard_handshake
    refresh: 10s
```

**Comment by droscy** [comment]
You are right, sorry, I should have updated esphome in venv. I was using old esphome code.

**Comment by droscy** [comment]
Code updated, these are the changes:

* increase wireguard setup priority
* add parameter `require_connection_to_proceed` (see [doc](https://deploy-preview-2948--esphome.netlify.app/components/wireguard))
* enforce validation for WireGuard keys
* use ESPTime from core esphome namespace

**Note** due to braking changes from #4926 and probably until next stable esphome release, you must use esphome from dev branch to compile wireguard.

**Comment by jesserockz** [comment]
I have not had a chance to review this PR yet sorry,but sorry about the breaking change for the time stuff.

In terms of the library, we only really adopt stuff in that core developers work on. There is no issue using the library under your username. You could always move it into its own org perhaps.

**Review by jesserockz — CHANGES_REQUESTED** [review]
Well, I just had a very quick look at the PR as I was curious why there are 3 individual components being added. 


This is not how it should be done btw. The status and handshake should just be binary_sensor and sensor platforms respectively inside the `wireguard` component.

So the yaml should look like this:

```yaml

wireguard:
  ...
  
sensor:
  - platform: wireguard
    handshake: 
      name: ...
    <future sensors exposed>

binary_sensor:
  - platform: wireguard
    status:
      name: Wireguard Status
    <future binary sensors exposed>
```

**Comment by droscy** [comment]
> This is not how it should be done btw. The status and handshake should just be binary_sensor and sensor platforms respectively inside the `wireguard` component.

Sure, I'll do. I was not aware of it and currently I have no idea on how to code such configuration but I'll check other components.

**Comment by jesserockz** [comment]
Mostly since the code is separate already, you can move the files from status and handshake into `binary_sensor` and `sensor` folders respectively inside the `wireguard` folder, rename the namespace and change the python filenames to `__init__.py`

**Comment by droscy** [comment]
Is a bad idea to automatically configure those sensors when the user configure the wireguard component?

**Comment by tomaszduda23** [comment]
> Is a bad idea to automatically configure those sensors when the user configure the wireguard component?

I was wondering exactly the same thing some time ago. Those are sually optional so you need to configure them explicitly. I was not able to find any examples that makes automatic configuration.

**Comment by droscy** [comment]
@jesserockz the sensors have been moved inside wireguard component itself.

```yaml
sensor:
  - platform: wireguard
    latest_handshake:
      name: 'WireGuard Latest Handshake'

binary_sensor:
  - platform: wireguard
    status:
      name: 'WireGuard Status'
```
I've named the latest handshake `latest_handshake` instead of `handshake`.

I think that there are some issues on GitHub side because the CI checks continue to fail but I cannot re-run them.

**Comment by CarlosGS** [comment]
I can confirm it works (with ESPHomes' dev branch), though it gives error "binary_sensor.h not found" if the binary_sensor component is not instantiated.

Edit: All working OK with latest changes, thanks! :tada:

**Comment by droscy** [comment]
Code updated, these are the changes:

* restart wireguard link after local network connection loss
* upgrade esp_wireguard to version 0.3.2
  - fix wrong latest handshake retrieval
  - add support for eth interface if wifi not available (thanks to @jpeletier)

**Comment by misuzu** [comment]
> * restart wireguard link after local network connection loss

Why does it needs restarting? Isn't wireguard stateless?

**Comment by droscy** [comment]
After a local connection loss, sometimes, the device could not reestablish the wireguard connection. I didn't find the cause, I suspect something related to keys' rotation in `esp_wireguard`, but really I don't know so, as a workaround, I decided to refresh the link.

If you want to debug such condition comment out `this->stop_connection_();` in method `Wireguard::loop()` of `wireguard.cpp` file.

**Comment by RoganDawes** [comment]
This worked perfectly, thank you!

If I could make a tiny documentation suggestion, to include the `peer_port:` configuration option in the example yaml that 99% of people will just copy and paste, so that people like me with non-default Wireguard port numbers won't have to wonder how to specify that port.

File: esphome/components/wireguard/binary_sensor.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        device_class=DEVICE_CLASS_CONNECTIVITY,
+    ),
+}
+
+
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_WIREGUARD_ID])
+
+    if CONF_STATUS:
+        sens = await binary_sensor.new_binary_sensor(config[CONF_STATUS])
```
Comment (jesserockz) [inline esphome/components/wireguard/binary_sensor.py]:
```suggestion
    if status_config := config.get(CONF_STATUS):
        sens = await binary_sensor.new_binary_sensor(status_config)
```

File: esphome/components/wireguard/sensor.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+    ),
+}
+
+
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_WIREGUARD_ID])
+
+    if CONF_LATEST_HANDSHAKE:
+        sens = await sensor.new_sensor(config[CONF_LATEST_HANDSHAKE])
```
Comment (jesserockz) [inline esphome/components/wireguard/sensor.py]:
```suggestion
    if latest_handshake_config := config.get(CONF_LATEST_HANDSHAKE):
        sens = await sensor.new_sensor(latest_handshake_config)
```

File: esphome/components/wireguard/wireguard.cpp [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    this->defer(std::bind(&Wireguard::start_connection_, this));  // defer to avoid blocking setup
+  } else {
+    ESP_LOGE(TAG, "cannot initialize WireGuard, error code %d", this->wg_initialized_);
+    this->mark_failed();
+  }
+}
+
+void Wireguard::loop() {
+  PollingComponent::loop();
+
```
Comment (jesserockz) [inline esphome/components/wireguard/wireguard.cpp]: Not required
```suggestion
```

File: esphome/components/wireguard/wireguard.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+// These are used for possibly long DNS resolution to temporarily suspend the watchdog
+void suspend_wdt();
+void resume_wdt();
+
+}  // namespace wireguard
+}  // namespace esphome
+
+#endif
+
+// vim: tabstop=2 shiftwidth=2 expandtab
```
Comment (jesserockz) [inline esphome/components/wireguard/wireguard.h]: Please don't add user specific settings to the files like this.
We have a `.editorconfig` file that exists in the root of the repo that should be used to set up your editor.
```suggestion
```

File: tests/test9.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  - platform: wireguard
+    status:
+      name: 'WireGuard Status'
+
+sensor:
+  - platform: wireguard
+    latest_handshake:
+      name: 'WireGuard Latest Handshake'
+
+# vim: ts=2 sw=2 et
```
Comment (jesserockz) [inline tests/test9.yaml]:
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Mostly looks good. Just a few minor changes below.

I dont have a wireguard setup myself so cannot test this.

**Comment by droscy** [comment]
> Mostly looks good. Just a few minor changes below.

Hi @jesserockz, changes applied.

**Review by jesserockz — DISMISSED** [review]

**Comment by jesserockz** [comment]
Just realised you might need to create a new test file

**Comment by droscy** [comment]
> Just realised you might need to create a new test file

I've just changed `9` to `10` in workflow file `ci.yml` in order to execute the renamed test file.

@jesserockz the `ci.yml` of the `dev` branch doesn't contain test files 9 and 9.1 but those files exist in `tests/` folder.

**Comment by jesserockz** [comment]
> > Just realised you might need to create a new test file
> 
> I've just changed `9` to `10` in workflow file `ci.yml` in order to execute the renamed test file.
> 
> @jesserockz the `ci.yml` of the `dev` branch doesn't contain test files 9 and 9.1 but those files exist in `tests/` folder.

#5346

File: tests/README.md [hunk 6]
```diff
@@ -27,3 +27,4 @@ Current test_.yaml file contents.
 | test6.yaml | RP2040 | wifi | N/A
 | test7.yaml | ESP32-C3 | wifi | N/A
 | test8.yaml | ESP32-S3 | wifi | None
+| test9.yaml | ESP32 | wifi | None
```
Comment (jesserockz) [inline tests/README.md]:
```suggestion
| test10.yaml | ESP32 | wifi | None
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**